enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,366 @@
+multiple microshift subpackages, supporting at least the following use cases:
+
+| Use case 	                                        | Packages installed in the RHEL4Edge image 	| Notes 	                            |
+|--------------------------------------------------	|-----------------------------------------  	|-------	                            |
+| The whole stack of components is needed         	| `microshift`, `microshift-networking-ovnk`, `microshift-storage-topolvm` | Being explicit about the components avoids unexpected changes to customers if the default components change; i.e. they would get an error if their component of choice got deprecated. |
+| Only advanced networking is needed                | `microshift`, `microshift-networking-ovnk`  |                                    	|
+| The bare minimum set of components is necessary   | `microshift`                             |                                   	|
+| Only persistent volumes are necessary             | `microshift`, `microshift-storage-topolvm` |                                     |
+
+
+The packages for networking and storage would be different in upstrean, containing
+references to unrestricted (OKD) container images, removing the need for a pull secret.
+
+The following changes would be introduced to allow the described type of modularity:
+
+Reuse and extend the kustomize loader so system and user components use the exact mechanism.
+
+`/usr/lib/microshift/components.d/004-multus/kustomize.yaml`
+`/usr/lib/microshift/components.d/005-networking-ovnk/kustomize.yaml`
+`/usr/lib/microshift/components.d/010-storage-topolvm/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory executing the component manifests in alphabetical order.
+
+  Today for the CNI/CSI components there is no wait or readiness conditions,
+  if this needed to be implemented in the future it could be defined in yaml
+  metadata.
+
+  `/etc/microshift/components.d/` would also be provided to allow passing down
+  specific configuration down to components; for example:
+
+  `/etc/microshift/components.d/010-storage-topolvm/config.yaml`
+
+   such yaml would parsed and passed down on the templates of the following
+   point as `{{ .config.value }}` or `{{ .config.sub.value }}`.
+
+*  Kustomize yamls would be pre-processed with go templates to allow insertion of
+  variables as we do today for core assets, i.e.: `{{ .KubeconfigDir }}`,
+  `{{ .BaseDomain }}`,
+  [etc...](https://github.com/openshift/microshift/blob/eb0488fd167333256c69f4eeedc5463c4035a532/pkg/components/render.go#L24).
+  Image references will be injected in the `images` section
+  of the kustomize format yaml.
+
+* The functionality used in the topo-lvm component to embed the lvmd config file
+  in a config map could be performed with a config map generator, which would contain
+  a default config file.
```

Comment:
Can kustomize inject the contents of a separate file into a template like that?

Reply:
There is a [configmap generator](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#configmapgenerator) which can take a reference to a file and inject it in a config map. The tricky part here will be that we need to figure out how to push a default, and if the user configures an explicit one, inject that one instead.

This specific part from CSI/topolvm is what I want to discuss with @copejon